packages/theme/src/cli/utilities/theme-environment/html.ts

```diff
@@ -0,0 +1,67 @@
+import {getProxyStorefrontHeaders, patchRenderingResponse} from './proxy.js'
+import {getInMemoryTemplates, injectHotReloadScript} from './hot-reload/server.js'
+import {render} from './storefront-renderer.js'
+import {defineEventHandler, sendWebResponse, type H3Error} from 'h3'
+import {renderError} from '@shopify/cli-kit/node/ui'
+import type {Theme} from '@shopify/cli-kit/node/themes/types'
+import type {DevServerContext} from './types.js'
+
+export function getHtmlHandler(theme: Theme, ctx: DevServerContext) {
+  return defineEventHandler(async (event) => {
+    const {path: urlPath, method, headers} = event
+
+    // eslint-disable-next-line no-console
+    console.log(`${method} ${urlPath}`)
+
+    const response = await render(ctx.session, {
+      path: urlPath,
+      query: [],
+      themeId: String(theme.id),
+      cookies: headers.get('cookie') || '',
+      sectionId: '',
+      headers: getProxyStorefrontHeaders(event),
+      replaceTemplates: getInMemoryTemplates(),
```

Comment:
I've been investigating why the render function is a bit slow; in some scenarios, it takes up to 10 seconds. It appears that the primary reason is that we're passing all JSON files here. The secondary reason is the lack of memoization in the renderer's authentication, which also impacts performance, but only by around 500ms.

I'm currently working on the memoization issue, so we don't need to address that here.

Regarding the passing of all JSON files, the call [here](https://github.com/Shopify/cli/blob/0610fc4e3da7fc3ecb19c6097285d85afc9e739e/packages/theme/src/cli/utilities/theme-environment/theme-environment.ts#L13) is synchronous, so in theory, we wouldn't need to pass all those files since everything is already uploaded.

However, we do want to pass some templates; we just need to ensure we're only passing those that haven't been uploaded yet.

Reply:
Ah great catch, will fix this! I was reading the JSON files early to have an in-memory version of the existing sections. I can just move that to a different object 👍